# PANDAS PRACTICAL ASSIGNMENT
GROUP NO:9

Participants:
1. Adili S. Edward
2. Joseph M. Nzije
3. Carol F. Kika
4. Frank L. Mlyakalamu
5. Athanas J. Kayombo

# SECTION A: CREATING PANDAS DATAFRAMES

A1. Creating a DataFrame from a List

In [ ]:
import pandas as pd

In [ ]:
# Create a list containing student records
data = [
    ['241001','ADILI', 14],
    ['241002','JOSEPH', 15],
    ['241003','CAROL', 13],
    ['241004','FRANK', 16],
    ['241005','ATHANAS', 18]
]

In [ ]:
# Convert the list into a DataFrame and Assigning appropriate column names
df = pd.DataFrame(data, columns=['RegNo:','Name', 'Age'])

In [ ]:
# Display the DataFrame
print("Student Records DataFrame:")
print(df)

Student Records DataFrame:
   RegNo:     Name  Age
0  241001    ADILI   14
1  241002   JOSEPH   15
2  241003    CAROL   13
3  241004    FRANK   16
4  241005  ATHANAS   18


This output shows that we successfully converted a Python list into a Pandas DataFrame. Each inner list represents one student, while the column names are Registration Number, Name, and Age.

A2. Creating a DataFrame from a Dictionary

In [ ]:
# Create a dictionary containing student information
data = {
    "RegNo": [241001, 241002, 241003, 241004, 241005],
    "Name": ["ADILI", "JOSEPH", "CAROL", "FRANK", "ATHANAS"],
    "Age": [14, 15, 13, 16, 18]
}

In [ ]:
# Convert the dictionary into a DataFrame
record = pd.DataFrame(data)

In [ ]:
# Display the DataFrame
print("Student Records Dictionary:")
display(record)

Student Records Dictionary:


,RegNo,Name,Age
0,241001,ADILI,14
1,241002,JOSEPH,15
2,241003,CAROL,13
3,241004,FRANK,16
4,241005,ATHANAS,18


Here, we created a DataFrame from a dictionary. The dictionary keys became the column names, while the values became the data in each column

In [ ]:
# Display the data types of each column in the DataFrame
record.dtypes

RegNo    int64
Name       str
Age      int64
dtype: object

This output shows the data type of each column. Registration number and age are integers, while the student's name is stored as a string.

A3. Creating a DataFrame from Series

In [ ]:
# Create three Pandas Series
student_names = pd.Series(["ADILI", "JOSEPH", "CAROL", "FRANK", "ATHANAS"])
student_ages = pd.Series([14, 15, 13, 16, 18])
student_scores = pd.Series([75, 82, 68, 91, 79])

In [ ]:
# Combine the Series into a DataFrame
students_from_series = pd.DataFrame({
    "Student Name": student_names,
    "Age": student_ages,
    "Score": student_scores
})

In [ ]:
# Display the DataFrame
print("DataFrame created from Series:")
display(students_from_series)

DataFrame created from Series:


,Student Name,Age,Score
0,ADILI,14,75
1,JOSEPH,15,82
2,CAROL,13,68
3,FRANK,16,91
4,ATHANAS,18,79


This output shows that we created three Pandas Series for names, ages, and scores and then combined them into one DataFrame. The result is a structured table containing information about each student.

# SECTION B: KAGGLE DATA ANALYSIS 
Dataset: Titanic Dataset

Kaggle - Titanic Dataset 

    Files: test.csv, train.csv

Kaggle Link https://www.kaggle.com/datasets/zain280/titanic-data-set

# 1. Importing the Dataset

In [ ]:
# Read the CSV files
train = pd.read_csv("Titanic-Dataset/train.csv")

In [ ]:
test = pd.read_csv("Titanic-Dataset/test.csv")

Training Data

In [ ]:
print("First 10 rows of Training Dataset:")
train.head(10)

First 10 rows of Training Dataset:


,PassengerId,Survived,Pclass,Name,Sex,Age,SibSp,Parch,Ticket,Fare,Cabin,Embarked
0,1,0,3,"Braund, Mr. Owen Harris",male,22.0,1,0,A/5 21171,7.2500,NaN,S
1,2,1,1,"Cumings, Mrs. John Bradley (Florence Briggs Th...",female,38.0,1,0,PC 17599,71.2833,C85,C
2,3,1,3,"Heikkinen, Miss. Laina",female,26.0,0,0,STON/O2. 3101282,7.9250,NaN,S
3,4,1,1,"Futrelle, Mrs. Jacques Heath (Lily May Peel)",female,35.0,1,0,113803,53.1000,C123,S
4,5,0,3,"Allen, Mr. William Henry",male,35.0,0,0,373450,8.0500,NaN,S
5,6,0,3,"Moran, Mr. James",male,NaN,0,0,330877,8.4583,NaN,Q
6,7,0,1,"McCarthy, Mr. Timothy J",male,54.0,0,0,17463,51.8625,E46,S
7,8,0,3,"Palsson, Master. Gosta Leonard",male,2.0,3,1,349909,21.0750,NaN,S
8,9,1,3,"Johnson, Mrs. Oscar W (Elisabeth Vilhelmina Berg)",female,27.0,0,2,347742,11.1333,NaN,S
9,10,1,2,"Nasser, Mrs. Nicholas (Adele Achem)",female,14.0,1,0,237736,30.0708,NaN,C


This output displays the first ten records from the Titanic training dataset. Each row represents one passenger, and the columns contain information such as survival status, passenger class, name, gender, age, fare, cabin, and port of embarkation.

In [ ]:
print("\nTraining Dataset Shape:")
train.shape


Training Dataset Shape:


(891, 12)

The training dataset contains 891 rows and 12 columns. Therefore, we have information about 891 passengers and 12 variables

In [ ]:
print("\nTraining Dataset Columns:")


Training Dataset Columns:


In [ ]:
train.columns

Index(['PassengerId', 'Survived', 'Pclass', 'Name', 'Sex', 'Age', 'SibSp',
       'Parch', 'Ticket', 'Fare', 'Cabin', 'Embarked'],
      dtype='str')

These are the 12 variables available in the training dataset. The Survived column is particularly important because it tells us whether each passenger survived or not.

In [ ]:
print("\nTraining Dataset Data Types:")



Training Dataset Data Types:


In [ ]:
train.dtypes

PassengerId      int64
Survived         int64
Pclass           int64
Name               str
Sex                str
Age            float64
SibSp            int64
Parch            int64
Ticket             str
Fare           float64
Cabin              str
Embarked           str
dtype: object

This output shows the data type of each variable. Numerical information such as PassengerId, Survived, Pclass, Age, and Fare is represented using numerical data types, while information such as Name and Sex is represented as text.

Testing Data

In [ ]:
print("\nFirst 10 rows of Testing Dataset:")


First 10 rows of Testing Dataset:


In [ ]:
test.head(10)


,PassengerId,Pclass,Name,Sex,Age,SibSp,Parch,Ticket,Fare,Cabin,Embarked
0,892,3,"Kelly, Mr. James",male,34.5,0,0,330911,7.8292,NaN,Q
1,893,3,"Wilkes, Mrs. James (Ellen Needs)",female,47.0,1,0,363272,7.0000,NaN,S
2,894,2,"Myles, Mr. Thomas Francis",male,62.0,0,0,240276,9.6875,NaN,Q
3,895,3,"Wirz, Mr. Albert",male,27.0,0,0,315154,8.6625,NaN,S
4,896,3,"Hirvonen, Mrs. Alexander (Helga E Lindqvist)",female,22.0,1,1,3101298,12.2875,NaN,S
5,897,3,"Svensson, Mr. Johan Cervin",male,14.0,0,0,7538,9.2250,NaN,S
6,898,3,"Connolly, Miss. Kate",female,30.0,0,0,330972,7.6292,NaN,Q
7,899,2,"Caldwell, Mr. Albert Francis",male,26.0,1,1,248738,29.0000,NaN,S
8,900,3,"Abrahim, Mrs. Joseph (Sophie Halaut Easu)",female,18.0,0,0,2657,7.2292,NaN,C
9,901,3,"Davies, Mr. John Samuel",male,21.0,2,0,A/4 48871,24.1500,NaN,S


This output shows the first ten records from the testing dataset. It has similar passenger information to the training dataset, but it does not contain the Survived column.

In [ ]:
print("\nTesting Dataset Shape:")


Testing Dataset Shape:


In [ ]:
test.shape

(418, 11)

The testing dataset contains 418 rows and 11 columns. It has fewer columns than the training dataset because it does not contain the Survived variable.

In [ ]:
print("\nTesting Dataset Columns:")


Testing Dataset Columns:


In [ ]:
test.columns

Index(['PassengerId', 'Pclass', 'Name', 'Sex', 'Age', 'SibSp', 'Parch',
       'Ticket', 'Fare', 'Cabin', 'Embarked'],
      dtype='str')

These are the 11 columns available in the testing dataset. The main difference from the training dataset is that Survived is missing.

In [ ]:
print("\nTesting Dataset Data Types:")


Testing Dataset Data Types:


In [ ]:
test.dtypes

PassengerId      int64
Pclass           int64
Name               str
Sex                str
Age            float64
SibSp            int64
Parch            int64
Ticket             str
Fare           float64
Cabin              str
Embarked           str
dtype: object

The output shows the data types of the testing dataset. The variables are correctly recognized as numerical or text data depending on their content.

# 2. Statistics

A. Average Passenger Age

In [ ]:
average_age = train["Age"].mean()

In [ ]:
print("Average passenger age:", average_age)

Average passenger age: 29.69911764705882


The average age of passengers in the training dataset was approximately 29.7 years. This means that, on average, the passengers were around 30 years old

B. Average Fare

In [ ]:
average_fare = train["Fare"].mean()

In [ ]:
print("Average fare:", average_fare)

Average fare: 32.204207968574636


The average fare paid by passengers was approximately 32.20. This represents the average ticket price in the dataset.

C. Maximum Fare

In [ ]:
maximum_fare = train["Fare"].max()

In [ ]:
print("Maximum fare:", maximum_fare)

Maximum fare: 512.3292


The maximum fare recorded in the dataset was approximately 512.33. This means that this was the highest ticket fare paid by a passenger in the training dataset.

D. Survival Rate

In [ ]:
survival_rate = train["Survived"].mean()

In [ ]:
print("Survival rate:", survival_rate)

Survival rate: 0.3838383838383838


The survival rate is approximately 0.384, which means that about 38.38 percent of the passengers survived

In [ ]:
print("Survival rate (%):", survival_rate * 100)

Survival rate (%): 38.38383838383838


When we convert the survival rate into a percentage, we get approximately 38.38 percent. Therefore, less than half of the passengers in the training dataset survived.

E. Descriptive Statistics

In [ ]:
# Training Dataset Summary Statistics
train.describe()

,PassengerId,Survived,Pclass,Age,SibSp,Parch,Fare
count,891.000000,891.000000,891.000000,714.000000,891.000000,891.000000,891.000000
mean,446.000000,0.383838,2.308642,29.699118,0.523008,0.381594,32.204208
std,257.353842,0.486592,0.836071,14.526497,1.102743,0.806057,49.693429
min,1.000000,0.000000,1.000000,0.420000,0.000000,0.000000,0.000000
25%,223.500000,0.000000,2.000000,20.125000,0.000000,0.000000,7.910400
50%,446.000000,0.000000,3.000000,28.000000,0.000000,0.000000,14.454200
75%,668.500000,1.000000,3.000000,38.000000,1.000000,0.000000,31.000000
max,891.000000,1.000000,3.000000,80.000000,8.000000,6.000000,512.329200


In [ ]:
#Testing Dataset Summary Statistics
test.describe()

,PassengerId,Pclass,Age,SibSp,Parch,Fare
count,418.000000,418.000000,332.000000,418.000000,418.000000,417.000000
mean,1100.500000,2.265550,30.272590,0.447368,0.392344,35.627188
std,120.810458,0.841838,14.181209,0.896760,0.981429,55.907576
min,892.000000,1.000000,0.170000,0.000000,0.000000,0.000000
25%,996.250000,1.000000,21.000000,0.000000,0.000000,7.895800
50%,1100.500000,3.000000,27.000000,0.000000,0.000000,14.454200
75%,1204.750000,3.000000,39.000000,1.000000,0.000000,31.500000
max,1309.000000,3.000000,76.000000,8.000000,9.000000,512.329200


This output gives us a statistical summary of the numerical variables. It helps us understand the distribution of passenger age, fare, passenger class, family relationships, and survival.

# 3. Data Cleaning

A. Check Missing Values

Training dataset

In [ ]:
print("Missing values in training dataset:")

Missing values in training dataset:


In [ ]:
train.isnull().sum()

PassengerId      0
Survived         0
Pclass           0
Name             0
Sex              0
Age            177
SibSp            0
Parch            0
Ticket           0
Fare             0
Cabin          687
Embarked         2
dtype: int64

Testing dataset

In [ ]:
print("Missing values in testing dataset:")

Missing values in testing dataset:


In [ ]:
test.isnull().sum()

PassengerId      0
Pclass           0
Name             0
Sex              0
Age             86
SibSp            0
Parch            0
Ticket           0
Fare             1
Cabin          327
Embarked         0
dtype: int64

Here we are checking the dataset for missing information. Missing values are important because they can affect our analysis and may cause errors when performing calculations

 B. Check Duplicates

In [ ]:
print("Duplicate rows in training:", train.duplicated().sum())

Duplicate rows in training: 0


In [ ]:
print("Duplicate rows in testing:", test.duplicated().sum())

Duplicate rows in testing: 0


We also checked for duplicate records. The result showed that there were no duplicate rows, so no duplicate records needed to be removed.

C. Handle Missing Ages

In [ ]:
train["Age"] = train["Age"].fillna(train["Age"].median())

Checking data consistency 

In [ ]:
train.isnull().sum()

PassengerId      0
Survived         0
Pclass           0
Name             0
Sex              0
Age              0
SibSp            0
Parch            0
Ticket           0
Fare             0
Cabin          687
Embarked         2
dtype: int64

This output confirms that the data-cleaning process was successful because every column now contains zero missing values.

In [ ]:
test["Age"] = test["Age"].fillna(train["Age"].median())

Checking data consistency

In [ ]:
test.isnull().sum()

PassengerId      0
Pclass           0
Name             0
Sex              0
Age              0
SibSp            0
Parch            0
Ticket           0
Fare             1
Cabin          327
Embarked         0
dtype: int64

This output confirms that the data-cleaning process was successful because every column now contains zero missing values.

D. Handle Missing Categorical Values

In [ ]:
train["Embarked"] = train["Embarked"].fillna(train["Embarked"].mode()[0])

In [ ]:
train["Cabin"] = train["Cabin"].fillna("Unknown")

Checking Data Consistency

In [ ]:
train.isnull().sum()

PassengerId    0
Survived       0
Pclass         0
Name           0
Sex            0
Age            0
SibSp          0
Parch          0
Ticket         0
Fare           0
Cabin          0
Embarked       0
dtype: int64

In [ ]:
test["Cabin"] = test["Cabin"].fillna("Unknown")

In [ ]:
test["Fare"] = test["Fare"].fillna(train["Fare"].median())

Checking Data Consistency

In [ ]:
test.isnull().sum()

PassengerId    0
Pclass         0
Name           0
Sex            0
Age            0
SibSp          0
Parch          0
Ticket         0
Fare           0
Cabin          0
Embarked       0
dtype: int64

After cleaning the data, the missing values were handled using appropriate methods. For example, missing numerical values such as age were replaced using the median, while missing categorical information was handled appropriately.

E. Convert Appropriate Variables to Categorical Data Types

In [ ]:
categorical_columns = ["Pclass", "Sex", "Embarked"]

In [ ]:
for column in categorical_columns:
    train[column] = train[column].astype("category")
    test[column] = test[column].astype("category")

In [ ]:
train["Survived"] = train["Survived"].astype("category")

We converted variables that represent categories into categorical data types. This makes the data more appropriate for grouping and categorical analysis.

# 4. Selecting Data

A. Female Passengers

In [ ]:
female_passengers = train[train["Sex"] == "female"]

In [ ]:
print(female_passengers)

     PassengerId Survived Pclass  \
1              2        1      1   
2              3        1      3   
3              4        1      1   
8              9        1      3   
9             10        1      2   
..           ...      ...    ...   
880          881        1      2   
882          883        0      3   
885          886        0      3   
887          888        1      1   
888          889        0      3   

                                                  Name     Sex   Age  SibSp  \
1    Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                               Heikkinen, Miss. Laina  female  26.0      0   
3         Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
8    Johnson, Mrs. Oscar W (Elisabeth Vilhelmina Berg)  female  27.0      0   
9                  Nasser, Mrs. Nicholas (Adele Achem)  female  14.0      1   
..                                                 ...     ...   ...    ...   
880       Shel

This output contains only female passengers. We selected them by filtering the Sex column where the value is female.

B. Male Passengers

In [ ]:
male_passengers = train[train["Sex"] == "male"]

In [ ]:
print(male_passengers)

     PassengerId Survived Pclass                            Name   Sex   Age  \
0              1        0      3         Braund, Mr. Owen Harris  male  22.0   
4              5        0      3        Allen, Mr. William Henry  male  35.0   
5              6        0      3                Moran, Mr. James  male  28.0   
6              7        0      1         McCarthy, Mr. Timothy J  male  54.0   
7              8        0      3  Palsson, Master. Gosta Leonard  male   2.0   
..           ...      ...    ...                             ...   ...   ...   
883          884        0      2   Banfield, Mr. Frederick James  male  28.0   
884          885        0      3          Sutehall, Mr. Henry Jr  male  25.0   
886          887        0      2           Montvila, Rev. Juozas  male  27.0   
889          890        1      1           Behr, Mr. Karl Howell  male  26.0   
890          891        0      3             Dooley, Mr. Patrick  male  32.0   

     SibSp  Parch            Ticket    

This output contains only male passengers. It demonstrates how Pandas can be used to filter a dataset based on a specific condition

C. Passengers in First Class

In [ ]:
first_class = train[train["Pclass"] == 1]

In [ ]:
print(first_class)

     PassengerId Survived Pclass  \
1              2        1      1   
3              4        1      1   
6              7        0      1   
11            12        1      1   
23            24        1      1   
..           ...      ...    ...   
871          872        1      1   
872          873        0      1   
879          880        1      1   
887          888        1      1   
889          890        1      1   

                                                  Name     Sex   Age  SibSp  \
1    Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
3         Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
6                              McCarthy, Mr. Timothy J    male  54.0      0   
11                            Bonnell, Miss. Elizabeth  female  58.0      0   
23                        Sloper, Mr. William Thompson    male  28.0      0   
..                                                 ...     ...   ...    ...   
871   Beckwith

This output contains passengers who travelled in First Class. We selected them by filtering the Pclass column where the value is 1.

D. Passengers Who Survived

In [ ]:
survived = train[train["Survived"] == 1]

In [ ]:
print(survived)

     PassengerId Survived Pclass  \
1              2        1      1   
2              3        1      3   
3              4        1      1   
8              9        1      3   
9             10        1      2   
..           ...      ...    ...   
875          876        1      3   
879          880        1      1   
880          881        1      2   
887          888        1      1   
889          890        1      1   

                                                  Name     Sex   Age  SibSp  \
1    Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                               Heikkinen, Miss. Laina  female  26.0      0   
3         Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
8    Johnson, Mrs. Oscar W (Elisabeth Vilhelmina Berg)  female  27.0      0   
9                  Nasser, Mrs. Nicholas (Adele Achem)  female  14.0      1   
..                                                 ...     ...   ...    ...   
875           

This output contains passengers whose Survived value is 1. In the Titanic dataset, a value of 1 means the passenger survived.

E. Passengers Who Paid Above-Average Fare

In [ ]:
average_fare = train["Fare"].mean()

In [ ]:
above_average_fare = train[train["Fare"] > average_fare]

In [ ]:
print(above_average_fare)

     PassengerId Survived Pclass  \
1              2        1      1   
3              4        1      1   
6              7        0      1   
23            24        1      1   
27            28        0      1   
..           ...      ...    ...   
856          857        1      1   
863          864        0      3   
867          868        0      1   
871          872        1      1   
879          880        1      1   

                                                  Name     Sex   Age  SibSp  \
1    Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
3         Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
6                              McCarthy, Mr. Timothy J    male  54.0      0   
23                        Sloper, Mr. William Thompson    male  28.0      0   
27                      Fortune, Mr. Charles Alexander    male  19.0      3   
..                                                 ...     ...   ...    ...   
856         Wi

This output shows passengers who paid more than the average fare of approximately 32.20. We first calculated the average fare and then selected passengers whose fare was greater than that value.

# 5. Grouping

A. Survival Rate by Gender

In [ ]:
train["Survived"] = train["Survived"].astype(int)

In [ ]:
survival_by_gender = train.groupby("Sex")["Survived"].mean()

In [ ]:
print("Survival Rate by Gender:")
print(survival_by_gender)

Survival Rate by Gender:
Sex
female    0.742038
male      0.188908
Name: Survived, dtype: float64


This output shows the survival rate according to gender. About 74.20 percent of female passengers survived, compared with only 18.89 percent of male passengers. Therefore, females had a much higher survival rate than males in this dataset.

B. Survival Rate by Passenger Class

In [ ]:
survival_by_class = train.groupby("Pclass")["Survived"].mean()

In [ ]:
print("Survival Rate by Passenger Class:")
print(survival_by_class)

Survival Rate by Passenger Class:
Pclass
1    0.629630
2    0.472826
3    0.242363
Name: Survived, dtype: float64


This output shows the survival rate according to passenger class. First-Class passengers had the highest survival rate at 62.96 percent, while Third-Class passengers had the lowest at 24.24 percent.

C. Average Fare by Passenger Class

In [ ]:
average_fare_class = train.groupby("Pclass", observed=True)["Fare"].mean()


In [ ]:
print(average_fare_class)

Pclass
1    84.154687
2    20.662183
3    13.675550
Name: Fare, dtype: float64


This output shows that First-Class passengers paid the highest average fare, approximately 84.15. Second-Class passengers paid about 20.66, while Third-Class passengers paid about 13.68.

D. Average Age by Gender

In [ ]:
average_age_gender = train.groupby("Sex", observed=True)["Age"].mean()

In [ ]:
print(average_age_gender)

Sex
female    27.929936
male      30.140676
Name: Age, dtype: float64


This output compares the average age of male and female passengers. Female passengers had an average age of approximately 27.93 years, while male passengers had an average age of approximately 30.14 years.

# 6. Combining Training and Testing Datasets

In [ ]:
combined = pd.concat(
    [train, test],
    ignore_index=True
)

In [ ]:
print("\n===== COMBINED DATASET =====")
print(combined.head())
print("Combined shape:", combined.shape)
print("Combined columns:")
print(combined.columns.tolist())


===== COMBINED DATASET =====
   PassengerId  Survived Pclass  \
0            1       0.0      3   
1            2       1.0      1   
2            3       1.0      3   
3            4       1.0      1   
4            5       0.0      3   

                                                Name     Sex   Age  SibSp  \
0                            Braund, Mr. Owen Harris    male  22.0      1   
1  Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                             Heikkinen, Miss. Laina  female  26.0      0   
3       Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                           Allen, Mr. William Henry    male  35.0      0   

   Parch            Ticket     Fare    Cabin Embarked  
0      0         A/5 21171   7.2500  Unknown        S  
1      0          PC 17599  71.2833      C85        C  
2      0  STON/O2. 3101282   7.9250  Unknown        S  
3      0            113803  53.1000     C123        S  
4      0        

After combining the two datasets, we obtained 1,309 passenger records. The Survived column is available for the training records, while it is not available for the testing records.

Why is pd.concat() Appropriate?

Finally, we combined the training and testing datasets using pd.concat(). This is appropriate because both datasets contain information about Titanic passengers and have a similar structure. The rows were combined vertically, and the index was reset using ignore_index=True.